test/lib.bash

```diff
@@ -115,12 +115,161 @@ function downstream_eventing_e2e_tests {
   return $failed
 }
 
+# == Upgrade testing
+
+function run_rolling_upgrade_tests {
+  logger.info "Running rolling upgrade tests"
+
+  local latest_cluster_version latest_serving_version latest_eventing_version \
+    rootdir scope serving_in_scope eventing_in_scope serving_prober_pid \
+    eventing_prober_pid prev_serving_version prev_eventing_version \
+    ocp_target_version
+
+  scope="${1:?Provide an upgrade scope as arg[1]}"
+  serving_in_scope="$(echo "${scope}" | grep -vq serving ; echo "$?")"
+  eventing_in_scope="$(echo "${scope}" | grep -vq eventing ; echo "$?")"
+
+  prev_serving_version="$(actual_serving_version)"
+  prev_eventing_version="$(actual_eventing_version)"
+
+  # Save the rootdir before changing dir
+  rootdir="$(dirname "$(dirname "$(realpath "${BASH_SOURCE[0]}")")")"
+
+  if (( eventing_in_scope )); then
+    prepare_knative_eventing_tests
+  fi
+  if (( serving_in_scope )); then
+    prepare_knative_serving_tests
+  fi
+
+  logger.info 'Testing with pre upgrade tests'
+
+  if (( serving_in_scope )); then
+    run_serving_preupgrade_test
+  fi
+  if (( eventing_in_scope )); then
+    run_eventing_preupgrade_test
+  fi
+
+  logger.info 'Starting prober tests'
+
+  if (( serving_in_scope )); then
+    start_serving_prober "${prev_serving_version}" /tmp/prober-pid
+    serving_prober_pid=$(cat /tmp/prober-pid)
+  fi
+  if (( eventing_in_scope )); then
+    start_eventing_prober /tmp/prober-pid
+    eventing_prober_pid=$(cat /tmp/prober-pid)
+  fi
+
+  if (( serving_in_scope )); then
+    wait_for_serving_prober_ready
+  fi
+  if (( eventing_in_scope )); then
+    wait_for_eventing_prober_ready
+  fi
+
+  if [[ $UPGRADE_SERVERLESS == true ]]; then
+    latest_serving_version="${KNATIVE_SERVING_VERSION/v/}"
+    latest_eventing_version="${KNATIVE_EVENTING_VERSION/v/}"
+
+    logger.info "Updating Serverless to ${CURRENT_CSV}"
+    logger.debug "Serving version: ${prev_serving_version} -> ${latest_serving_version}"
+    logger.debug "Eventing version: ${prev_eventing_version} -> ${latest_eventing_version}"
+
+    approve_csv "$CURRENT_CSV" "$OLM_UPGRADE_CHANNEL"
+    if (( serving_in_scope )); then
+      check_serving_upgraded "${latest_serving_version}"
+    fi
+    if (( eventing_in_scope )); then
+      check_eventing_upgraded "${latest_eventing_version}"
```

Comment:
Where is this `check_eventing_upgraded` function defined? I can't find it anywhere.

Reply:
It wasn't. That's why I expected that those tests will fail.